File: app/src/main/java/ru/test/weatherapp/MainActivity.java
package ru.test.weatherapp;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "tagON";
    private String messageText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        final TextView temperValue = findViewById(R.id.temperValue);
        final TextView wetness = findViewById(R.id.text_wetness_content_main);
        final TextView windSpeed = findViewById(R.id.text_wind_speed_content_main);
        final TextView airPressure = findViewById(R.id.text_air_pressure_content_main);
        final TextView textCity = findViewById(R.id.text_city_content_main);

        textCity.setText(Settings.instance().getCity());

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });

        messageText = ((savedInstanceState == null) ? "Первый запуск!" : "Повторный запуск!") + " - onCreate()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
        Button button = findViewById(R.id.button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getWeather(temperValue, wetness, windSpeed, airPressure);
            }
        });
        getWeather(temperValue, wetness, windSpeed, airPressure);
    }

    private void getWeather(TextView temperValue,
                            TextView wetness, TextView windSpeed, TextView airPressure) {
        int temperature = (int) (Math.random() * 30);
        String temperText = ((Math.random() > 0.5) ? "" : "-") + temperature + " \u00B0C";
        temperValue.setText(temperText);

        wetness.setVisibility(View.INVISIBLE);
        windSpeed.setVisibility(View.INVISIBLE);
        airPressure.setVisibility(View.INVISIBLE);

        if (Settings.instance().getAirPressure()) {
            airPressure.setVisibility(View.VISIBLE);
            String airPressureText = "Давление воздуха: " + ((int) (Math.random() * 1000)) + " мм рт. с.";
            airPressure.setText(airPressureText);
        }

        if (Settings.instance().getWetness()) {
            wetness.setVisibility(View.VISIBLE);
            String wetnessText = "Влажность: " + ((int) (Math.random() * 100)) + " %";
            wetness.setText(wetnessText);
        }

        if (Settings.instance().getWindSpeed()) {
            windSpeed.setVisibility(View.VISIBLE);
            String windSpeedText = "Скорость ветра: " + ((int) (Math.random() * 20)) + " м/с";
            windSpeed.setText(windSpeedText);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    protected void onStart() {
        super.onStart();
        messageText = "onStart()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }

    @Override
    protected void onRestoreInstanceState(Bundle saveInstanceState) {
        super.onRestoreInstanceState(saveInstanceState);
        messageText = "Повторный запуск!! - onRestoreInstanceState()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }

    @Override
    protected void onResume() {
        super.onResume();
        messageText = "onResume()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }

    @Override
    protected void onPause() {
        super.onPause();
        messageText = "onPause()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }

    @Override
    protected void onSaveInstanceState(Bundle saveInstanceState) {
        super.onSaveInstanceState(saveInstanceState);
        messageText = "onSaveInstanceState()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }

    @Override
    protected void onStop() {
        super.onStop();
        messageText = "onStop()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        messageText = "onRestart()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        messageText = "onDestroy()";
        Toast.makeText(getApplicationContext(), messageText, Toast.LENGTH_SHORT).show();
        Log.d(TAG, messageText);
    }
}

File: app/src/main/java/ru/test/weatherapp/MainActivity.kt
package ru.test.weatherapp

import android.content.Context
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*


class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener, SettingsFragment.OnFragmentInteractionListener, ContentFragment.OnFragmentInteractionListener, HistoryFragment.OnFragmentInteractionListener {
    val settingsFragment = SettingsFragment()
    val contentFragment = ContentFragment()
    val historyFragment = HistoryFragment()
    private var currentFragment: Int = 1
    private var previousFragment: Int = 1

    override fun onFragmentInteraction(currentFragment: Int, nextFragment: Int) {
        changeFragment(currentFragment, nextFragment)
    }

    private fun changeFragment(currentFragment: Int, nextFragment: Int) {
        this.previousFragment = currentFragment
        this.currentFragment = nextFragment
        var fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.content_main_frame, getCurrentFragment(nextFragment))
        fragmentTransaction.commit()
    }

    private fun getCurrentFragment(fragmentNumber: Int): Fragment {
        return when (fragmentNumber) {
            1 -> settingsFragment
            2 -> contentFragment
            3 -> historyFragment
            else -> settingsFragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        fab.setOnClickListener { view ->
            Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show()
        }

        val toggle = ActionBarDrawerToggle(
                this, drawer_layout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer_layout.addDrawerListener(toggle)
        toggle.syncState()

        nav_view.setNavigationItemSelectedListener(this)

        var fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.add(R.id.content_main_frame, settingsFragment)
        fragmentTransaction.commit()


    }

    override fun onBackPressed() {
        if (drawer_layout.isDrawerOpen(GravityCompat.START)) {
            drawer_layout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        item.isChecked = true
        when (item.itemId) {
            R.id.about_dev -> {

            }
            R.id.write_me -> {

            }
        }
        drawer_layout.closeDrawer(GravityCompat.START)
        return true
    }


    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        return if (keyCode == KeyEvent.KEYCODE_BACK && currentFragment != 1) {
            changeFragment(previousFragment - 1, previousFragment)
            true
        } else {
            super.onKeyDown(keyCode, event)
        }
    }

    override fun onDestroy() {
        val editor = getPreferences(Context.MODE_PRIVATE).edit()
        editor.putString("cityName", Settings.instance().city)
        editor.putBoolean("wetness", Settings.instance().wetness)
        editor.putBoolean("airPressure", Settings.instance().airPressure)
        editor.putBoolean("windSpeed", Settings.instance().windSpeed)
        editor.apply()
        super.onDestroy()
    }
}
